### PR TITLE
IGNITE-16631 ODBC: Add ability to set query engine

### PR DESCRIPTION
**Pull request by ivandasch**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Review by alex-plekhanov — APPROVED** [review]

File: modules/platforms/cpp/odbc-test/src/cross_engine_test.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+        ret = SQLFetch(stmt);
+
+        if (!SQL_SUCCEEDED(ret))
+            BOOST_FAIL(GetOdbcErrorMessage(SQL_HANDLE_STMT, stmt));
+
+        ret = SQLFreeStmt(stmt, SQL_CLOSE);
+        ODBC_FAIL_ON_ERROR(ret, SQL_HANDLE_STMT, stmt);
+
+        return ToString(strBuf, strLen);
```
Comment (isapego) [inline modules/platforms/cpp/odbc-test/src/cross_engine_test.cpp:101]: `utility::SqlStringToString()`?

File: modules/platforms/cpp/odbc/src/config/configuration.cpp [hunk 2]
```diff
@@ -513,6 +529,14 @@ namespace ignite
                 if (value.IsSet())
                     map[key] = NestedTxMode::ToString(value.GetValue());
             }
+
+            template<>
+            void Configuration::AddToMap(ArgumentMap& map, const std::string& key,
+                                         const SettableValue<EngineMode::Type>& value)
```
Comment (isapego) [inline modules/platforms/cpp/odbc/src/config/configuration.cpp:535]: Formatting

File: modules/platforms/cpp/odbc/src/message.cpp [hunk 3]
```diff
@@ -88,6 +88,16 @@ namespace ignite
 
             if (version >= ProtocolVersion::VERSION_2_7_0)
                 writer.WriteInt8(config.GetNestedTxMode());
+
+            if (version >= ProtocolVersion::VERSION_2_13_0)
+            {
+                EngineMode::Type mode = config.GetEngineMode();
+
+                if (mode == EngineMode::DEFAULT)
+                    writer.WriteString(0, 0);
```
Comment (isapego) [inline modules/platforms/cpp/odbc/src/message.cpp:97]: Maybe write null here?

**Review by isapego — CHANGES_REQUESTED** [review]
C++ part looks good overall to me. Please see my comments inline.

**Review by isapego — APPROVED** [review]

File: modules/platforms/cpp/odbc-test/src/cross_engine_test.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+        oss << "QUERY_ENGINE=" << EngineMode::ToString(EType);
+
+        return oss.str();
+    }
+
+    template<EngineMode::Type EType>
+    std::string GetQueryEngine() {
+        SQLRETURN ret = ExecQuery(EType == EngineMode::CALCITE ? "select query_engine()"
+                                                               : "select public.query_engine()");
```
[the remark this answers is not in the document]
Comment (ivandasch) [reply modules/platforms/cpp/odbc-test/src/cross_engine_test.cpp:84]: Unfortunately, this is the case of current implementation of this function in `sql-calcite` branch. 
I suppose that this quirk should be a topic of other issue. WDYT?

**Pull request merged**